test/utils/setupKarma.js

```diff
@@ -0,0 +1,51 @@
+/* eslint-env mocha */
+import { createMochaHooks } from './mochaHooks';
+
+const mochaHooks = createMochaHooks(window.Mocha);
+
+before(function beforeAllHook() {
+  mochaHooks.beforeAll.forEach((mochaHook) => {
+    mochaHook.call(this);
+  });
+});
+
+after(function afterAllHook() {
+  mochaHooks.afterAll.forEach((mochaHook) => {
+    mochaHook.call(this);
+  });
+});
+
+beforeEach(function beforeEachHook() {
+  mochaHooks.beforeEach.forEach((mochaHook) => {
+    mochaHook.call(this);
+  });
+});
+
+afterEach(function afterEachHook() {
+  mochaHooks.afterEach.forEach((mochaHook) => {
+    mochaHook.call(this);
+  });
+});
+
+// Ensure that uncaught exceptions between tests result in the tests failing.
```

Comment:
https://github.com/mui-org/material-ui-x/pull/543 was meant to be the reproduction. It fails on this test https://github.com/mui-org/material-ui-x/blob/453e2f14f6981cda278faf0ce1edb8a7e543078d/packages/grid/x-grid/src/XGrid.test.tsx#L80-L89 (and the others). If you want to reproduce with fewer changes you can do:

- Check out a commit before https://github.com/mui-org/material-ui-x/pull/543
- Change the logic of the component to throw an async error:
```diff
diff --git a/packages/grid/_modules_/grid/hooks/utils/useResizeContainer.ts b/packages/grid/_modules_/grid/hooks/utils/useResizeContainer.ts
index e9750e8a..4297f4c1 100644
--- a/packages/grid/_modules_/grid/hooks/utils/useResizeContainer.ts
+++ b/packages/grid/_modules_/grid/hooks/utils/useResizeContainer.ts
@@ -8,6 +8,7 @@ export function useResizeContainer(apiRef): (size: ElementSize) => void {

   const onResize = React.useCallback(
     (size: ElementSize) => {
+      throw new Error('ooop')
       if (size.height === 0) {
         gridLogger.warn(
           [
diff --git a/test/karma.conf.js b/test/karma.conf.js
```
- Run the test `yarn test:karma`
- 🔥 the first test pass, all the other are skipped

Reply:
So just to be clear:
1. This does not cover all cases.
1. What we should catch is components that don't clean up.
   The problem isn't that an error was thrown between tests but that the component/test leaks.

I'm not that happy with the new listener since it might trick people into thinking errors between tests are caught and you're safe. Let's see if that produces problematic component code/tests. In the end we shouldn't actually add code we don't need.

Reply:
What do you recommend we do here? No strong preferences. If I can get `test/utils/setupKarma.js` to be shared with X,   while needed to keep the patch for https://github.com/karma-runner/karma-mocha/issues/227 on X, it will already be a good step forward in sharing code.

Reply:
Nah, all good to merge. Just want to caution that we don't get a false sense of security.